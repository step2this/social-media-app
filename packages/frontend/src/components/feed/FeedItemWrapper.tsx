import React from 'react';
import type { PostWithAuthor } from '@social-media-app/shared';
import { PostCard } from '../posts/PostCard';
import { useFeedItemAutoRead } from '../../hooks/useFeedItemAutoRead';

/**
 * Props for FeedItemWrapper component
 * Uses TypeScript best practices with explicit property types
 */
export interface FeedItemWrapperProps {
  /** Post data including author information */
  readonly post: PostWithAuthor;
  /** Whether to display the post in compact mode (default: true) */
  readonly compact?: boolean;
}

/**
 * Wrapper component for PostCard with auto-read functionality
 * 
 * Handles Instagram-like behavior where posts are marked as read after viewing.
 * Uses intersection observer via useFeedItemAutoRead hook to track visibility.
 * 
 * @example
 * ```tsx
 * <FeedItemWrapper post={post} compact={true} />
 * ```
 * 
 * TypeScript patterns applied from SKILL.md:
 * - Readonly modifier for immutable props
 * - Explicit interface for better error messages
 * - Optional properties with default values
 * - JSDoc comments for documentation
 */
export const FeedItemWrapper: React.FC<FeedItemWrapperProps> = ({ 
  post, 
  compact = true 
}) => {
  const elementRef = useFeedItemAutoRead(post.id);

  return (
    <div ref={elementRef as React.RefObject<HTMLDivElement>} className="feed-item-wrapper" data-testid="feed-item-wrapper">
      <PostCard post={post} compact={compact} />
    </div>
  );
};
