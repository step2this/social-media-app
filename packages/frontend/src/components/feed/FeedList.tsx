import React from 'react';
import type { PostWithAuthor } from '@social-media-app/shared';
import { FeedItemWrapper } from './FeedItemWrapper';

/**
 * Props for FeedList component
 * 
 * TypeScript patterns from SKILL.md:
 * - Readonly modifier for immutable collections
 * - ReadonlyArray ensures posts cannot be mutated
 */
export interface FeedListProps {
  /** Array of posts to display in the feed */
  readonly posts: readonly PostWithAuthor[];
  /** Whether to display posts in compact mode (default: true) */
  readonly compact?: boolean;
}

/**
 * Feed list component that renders a collection of posts
 * 
 * Uses TypeScript advanced patterns:
 * - Readonly arrays for immutability
 * - Type-safe iteration with proper key handling
 * - Generic-free design (uses concrete types for clarity)
 * 
 * @example
 * ```tsx
 * <FeedList posts={posts} compact={true} />
 * ```
 * 
 * TypeScript patterns applied from SKILL.md:
 * - Readonly modifiers prevent unintended mutations
 * - Explicit interfaces for better error messages
 * - JSDoc comments for documentation
 * - Const assertions where appropriate
 */
export const FeedList: React.FC<FeedListProps> = ({ 
  posts, 
  compact = true 
}) => {
  return (
    <div className="feed-list" data-testid="feed-list">
      {posts.map((post) => (
        <FeedItemWrapper key={post.id} post={post} compact={compact} />
      ))}
    </div>
  );
};
