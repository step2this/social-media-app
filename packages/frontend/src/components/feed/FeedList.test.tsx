import { describe, it, expect, vi, afterEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import type { PostWithAuthor } from '@social-media-app/shared';
import { FeedList } from './FeedList';
import { FeedItemWrapper } from './FeedItemWrapper';

// Mock FeedItemWrapper
vi.mock('./FeedItemWrapper', () => ({
  FeedItemWrapper: vi.fn(({ post }) => (
    <div data-testid={`feed-item-${post.id}`}>Feed Item {post.id}</div>
  ))
}));

describe('FeedList', () => {
  const mockPosts: PostWithAuthor[] = [
    {
      id: 'post-1',
      userId: 'user-1',
      userHandle: 'user1',
      imageUrl: 'https://example.com/image1.jpg',
      caption: 'First post',
      likesCount: 0,
      commentsCount: 0,
      createdAt: '2024-01-01T00:00:00.000Z',
      authorId: 'user-1',
      authorHandle: 'user1',
      authorFullName: 'User One',
    },
    {
      id: 'post-2',
      userId: 'user-2',
      userHandle: 'user2',
      imageUrl: 'https://example.com/image2.jpg',
      caption: 'Second post',
      likesCount: 0,
      commentsCount: 0,
      createdAt: '2024-01-02T00:00:00.000Z',
      authorId: 'user-2',
      authorHandle: 'user2',
      authorFullName: 'User Two',
    }
  ];

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders empty list', () => {
    render(<FeedList posts={[]} />);
    const list = screen.getByTestId('feed-list');
    expect(list).toBeInTheDocument();
    expect(list.children).toHaveLength(0);
  });

  it('renders single post', () => {
    render(<FeedList posts={[mockPosts[0]]} />);
    expect(screen.getByTestId('feed-item-post-1')).toBeInTheDocument();
  });

  it('renders multiple posts', () => {
    render(<FeedList posts={mockPosts} />);
    expect(screen.getByTestId('feed-item-post-1')).toBeInTheDocument();
    expect(screen.getByTestId('feed-item-post-2')).toBeInTheDocument();
  });

  it('each post has FeedItemWrapper', () => {
    render(<FeedList posts={mockPosts} />);
    expect(FeedItemWrapper).toHaveBeenCalledTimes(2);
  });

  it('passes compact prop to items', () => {
    render(<FeedList posts={[mockPosts[0]]} compact={false} />);
    expect(FeedItemWrapper).toHaveBeenCalledWith(
      expect.objectContaining({
        post: mockPosts[0],
        compact: false
      }),
      expect.anything()
    );
  });

  it('uses post.id as key', () => {
    const { container } = render(<FeedList posts={mockPosts} />);
    const items = container.querySelectorAll('[data-testid^="feed-item-"]');
    expect(items[0]).toHaveAttribute('data-testid', 'feed-item-post-1');
    expect(items[1]).toHaveAttribute('data-testid', 'feed-item-post-2');
  });

  it('has feed-list class', () => {
    const { container } = render(<FeedList posts={mockPosts} />);
    const list = container.querySelector('.feed-list');
    expect(list).toBeInTheDocument();
  });

  it('matches snapshot (empty)', () => {
    const { container } = render(<FeedList posts={[]} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot (with posts)', () => {
    const { container } = render(<FeedList posts={mockPosts} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const { container } = render(<FeedList posts={mockPosts} />);
    expect(container).toBeInTheDocument();
  });
});
