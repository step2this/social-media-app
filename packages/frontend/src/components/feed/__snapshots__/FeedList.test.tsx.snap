// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FeedList > matches snapshot (empty) 1`] = `
<div
  class="feed-list"
  data-testid="feed-list"
/>
`;

exports[`FeedList > matches snapshot (with posts) 1`] = `
<div
  class="feed-list"
  data-testid="feed-list"
>
  <div
    data-testid="feed-item-post-1"
  >
    Feed Item 
    post-1
  </div>
  <div
    data-testid="feed-item-post-2"
  >
    Feed Item 
    post-2
  </div>
</div>
`;
